### PR TITLE
Introduce sclang (interpreter) lexer, parser, and compiler regression tests

### PR DESCRIPTION
**Pull request by brianlheim**
You may want to see [this gist](https://gist.github.com/brianlheim/3ae1c2771f06ce653b35aad017b89d6b) for my original proposal.

~Also, this PR is apparently 1.4 mil new lines... sorry not sorry. :)~

I'm going with a zip archive. These tests have to be done manually anyway.

# Purpose

This is a major insurance PR. I want to make some pretty significant changes to the lexer/parser to solve some longstanding issues in SC, and obviously don't want to be doing it in the dark. This provides regression tests and related utilities that will allow us to examine the behavior of the lexer/parser/compiler under a wide and thorough (in some cases, complete) set of conditions. It will also be useful for anyone else who wants to make changes to the language in the future. These tests may seem excessive, but actually they are excessive, and that's OK.

# Summary of additions

The main additions are (1) `LexerParserCompilerTestUtils.sc`, which provides the bulk of test code, file formatting, and IO functionality; and (2) `TestLPCTestUtils.sc`, which provides some basic functionality tests for (1). Whether or not that stays in this repository is up to you, although I like having tests (obviously).

The rest of the files are either test files or validation files. The test files are short bits of code that actually run the tests, and can also be used to generate the validation files. The validation files are the real regression data that future changes will be tested against. The test utilities provide the means to clearly display the list of differences, additions, and subtractions between test and validation outputs.

# Test format

Generally, the operation is formalized as a series of `interpret`ations or `bytecode`ifications of "all possible strings", where each string is made of a sequence, of fixed length, of symbols from a predefined alphabet. For a given all-possible-strings test, an optional prefix and/or suffix may be provided.

For instance, if the alphabet is all lowercase letters [a-z], the prefix "hello_", the suffix "_world", the sequence length 3, then the tests would cover:
```
hello_aaa_world
hello_aab_world
hello_aac_world
...
hello_mqx_world
...
hello_zzy_world
hello_zzz_world
```

Generally, two methods are used:
* Compile the string. If that succeeds, interpret it. Record the resulting object using `.asString`, and if applicable, the resulting object's class.
* Compile the string. If that succeeds, record its FunctionDef's bytecodes.

If any step fails, a unique ID indicating that either a compile-time or run-time error occurred is recorded. In this way, we can compare the output of the parser/lexer/compiler across versions of SC to see what input is legal, what is illegal, and how certain expressions compile (or don't compile).

# Summary of tests

* Lexer:
   * strings up to length 3
   * prefix/suffix combos for line and block comments, { }, code bits added before and after
* Parser:
   * strings up to length 8 with an alphabet consisting of parser-important things like "var", "arg", ";"
   * a prefix/suffix combo for { }

See test files for full details, or I could add a .md file to this PR to explain what I've already explained here.

# TODO

- [x] Thorough review of .sc files
- [x] Is this the right folder? I don't want to run these tests every time in Travis; they take almost an hour, and rightly so. They should really only apply when deeper layers of sclang have been modified.
- [x] Validate across platforms ~(unlikely to conflict)~ (I was so naïve)
- [x] Validate across locales (possible conflicts)
- [x] Add tests targeting specific subroutines of the lexer (accidentals, floats, radix/hex notation) - that's why this is a work-in-progress
- [x] Add tests that target the compiler by repurposing the lexer tests
- [x] Finish documentation via readme
- [x] Add notes elsewhere about the existence of these tests

### CODE REVIEW

**Comment by vivid-synth** [comment]
Daaaaaaaang @brianlheim , awesome! Haven't looked in depth yet but very impressed

**Comment by snappizz** [comment]
!!!!!

**Comment by muellmusik** [comment]
Will try to have a look, though this is a bit outside my expertise! But in any case, thanks!!

**Comment by brianlheim** [comment]
Thanks guys :)

# Example use case

Let's say we have our original alphabet, `A, B, C`. Then at some point in the future, we also want to be able to ask questions, so we add `?`. (Note: this is a very real situation, because at some point we're going to add `^` back into the interpreter character set. Right now it causes unwanted bailing)

Let's also simulate a grammar change by trying out the same input with each string prefixed by a comment string. Maybe we decided that everything that doesn't compile should produce `nil` now.

```supercollider
(
var alphabet1 = ["A", "B", "C"];
var alphabet2 = ["A", "B", "C", "?"];

// emit the "correct" file
LexerParserCompilerTestUtils.testAllPossibleStrings(
	alphabet1,         // alphabet
	2,                 // string length
	"",                // prefix
	"",                // suffix
	"acceptancetest1", // test name
	\compile,          // technique to use
	true               // 'true' to emit validation files
);

// the alphabet has expanded—we should see new entries as a result
~diffs = LexerParserCompilerTestUtils.testAllPossibleStrings(
	alphabet2,
	2,
	"",
	"",
	"acceptancetest1",
	\compile,
	false
);

"\ndifferent alphabet\n".toUpper.postln;

LexerParserCompilerTestUtils.printDiffs(~diffs);

// precede every string with a comment—we should see `nil`s instead of compiling errors
~diffs = LexerParserCompilerTestUtils.testAllPossibleStrings(
	alphabet1,
	2,
	"//",
	"",
	"acceptancetest1",
	\compile,
	false
);

"\ndifferent prefix\n".toUpper.postln;

LexerParserCompilerTestUtils.printDiffs(~diffs);
)
```

# Results

I'm omitting the parser error messages for clarity.

```
/* warnings, status messages, parser error messages */

DIFFERENT ALPHABET

0 entries were missing from file 1
----------------------------------

7 entries were missing from file 2
----------------------------------
??: compile-error vs (no entry)
?A: compile-error vs (no entry)
?B: compile-error vs (no entry)
?C: compile-error vs (no entry)
A?: compile-error vs (no entry)
B?: compile-error vs (no entry)
C?: compile-error vs (no entry)

0 entries were different between files
--------------------------------------

/* warnings, status messages */

DIFFERENT PREFIX

0 entries were missing from file 1
----------------------------------

0 entries were missing from file 2
----------------------------------

9 entries were different between files
--------------------------------------
AA: nil:Nil vs compile-error
AB: nil:Nil vs compile-error
AC: nil:Nil vs compile-error
BA: nil:Nil vs compile-error
BB: nil:Nil vs compile-error
BC: nil:Nil vs compile-error
CA: nil:Nil vs compile-error
CB: nil:Nil vs compile-error
CC: nil:Nil vs compile-error
```

Let's also imagine that at some point in the future we decide we only need two symbols in our alphabet:

```supercollider
var alphabet3 = ["A", "B"];
~diffs = LexerParserCompilerTestUtils.testAllPossibleStrings(
	alphabet3,
	2,
	"",
	"",
	"acceptancetest1",
	\compile,
	false
);

"\nsmaller alphabet\n".toUpper.postln;

LexerParserCompilerTestUtils.printDiffs(~diffs);
```

output:
```
/* warnings and status messages */

SMALLER ALPHABET

5 entries were missing from file 1
----------------------------------
AC: (no entry) vs compile-error
BC: (no entry) vs compile-error
CA: (no entry) vs compile-error
CB: (no entry) vs compile-error
CC: (no entry) vs compile-error

0 entries were missing from file 2
----------------------------------

0 entries were different between files
--------------------------------------
```

**Comment by brianlheim** [comment]
Oh also, if you're going to try this out, I'd strongly recommend using the "silent SC" branch I've got here: https://github.com/brianlheim/supercollider/tree/special-silent-sc

It's exactly like master-dev, it just doesn't print error messages. Otherwise, you'll basically be getting something on the order of 200 billion lines of parser error output when you try to run the full suite.

File: testsuite/sclang/TestLexerBrutal.sc [hunk 1]
```diff
@@ -19,68 +16,37 @@ TestLexerBrutal : UnitTest {
...
+			alphabets = alphabets.collect {
 				arg alphabet;
 				alphabet = alphabet.reject(_==caretAscii);
 			}
 		};
 
-		alphabets.do {
+		alphabets = alphabets.collect {
 			arg alphabet;
 			alphabet = alphabet.collect(_.asAscii).collect(_.asString);
```
Comment (telephon) [inline testsuite/sclang/TestLexerBrutal.sc]: better omit the assignment (`alphabet = `):
 `alphabet.collect(_.asAscii).collect(_.asString);`

Comment (brianlheim) [inline testsuite/sclang/TestLexerBrutal.sc]: Currently in the process of refactoring this!

Comment (telephon) [inline testsuite/sclang/TestLexerBrutal.sc]: ah goo. I hope these sorts of comments are helpful at all? I'm never sure if I should make them.

Comment (brianlheim) [inline testsuite/sclang/TestLexerBrutal.sc]: Yes, they're very helpful! :) Sorry, I should have written that I was still planning to make major refactors. I'm in the process of giving the TestUtils files one more pass.

Comment (telephon) [inline testsuite/sclang/TestLexerBrutal.sc]: my pleasure! Just wanted to be sure :) Thanks for these thorough test cases, looking forward to more flexibility in the future!

File: testsuite/sclang/LexerParserCompilerTestUtils.sc [hunk 2]
```diff
@@ -22,138 +22,127 @@ LexerParserCompilerTestUtils {
...
 				postln("testAllPossibleStrings: Deleting test file");
-				File.delete(filename);
+				File.delete(file.pathName);
 			}
-		};
-
-		// postln("testAllPossibleStrings: Success!");
-
-		^diffs;
+		} {
```
Comment (telephon) [inline testsuite/sclang/LexerParserCompilerTestUtils.sc]: no need for the semicolon - better to leave it out for retturn val

**Comment by brianlheim** [comment]
Something I forgot—compiling to bytecode will return different results depending on the setting of `Process.tailCallOptimize`. For the sake of thoroughness I'm going to double the parser tests to take this into account. The compiler tests will be structured in two parts:
* use an alphabet like lexer's `half` alphabet (only printing chars + space) and compile to bytecode
* use an alphabet and small string length targeting special bytecode emission (for instance, messages like `.neg`, `.not`, and equality operators have unique opcodes)

File: testsuite/sclang/TestLexerTargetedBrutal.sc [hunk 3]
```diff
@@ -14,7 +14,7 @@ TestLexerTargetedBrutal : AbstractBrutalTest {
 		alphanumAlphabet = alphanumAlphabet.collect(_.asString);
 		alphanumAlphabet = alphanumAlphabet ++ ["pi", "+", "-", "."];
 
-		alphabets = [
+		alphabets = Dictionary[
```
Comment (telephon) [inline testsuite/sclang/TestLexerTargetedBrutal.sc]: IdentityDictionary will be faster (if that matters at all)

**Comment by brianlheim** [comment]
I'm done implementing everything I intended to now! Ready for thorough code review. As always, I welcome feedback, and I especially welcome it on naming and organization.

## Real-world example

I checked out `master` from a couple weeks ago, before #2625 (fix behavior of lexer with strings like `/* /* */*/`) was merged. I also merged in what I could from my `special-silent-sc` branch and ran the full suite of tests ("full suite" meaning the state of this branch at https://github.com/supercollider/supercollider/pull/2751/commits/1f2c0208512e2f6eb82417e07cef68f8385ad4fc). Here is the one diff file that was generated:

`mini_4_blockCommentEncloseSemanticSuffix_diff`

The rest of the tests passed. For context, keep in mind that each string below is tested with the prefix `/*` and `*/\unique`. In plain English, a string that—when inserted between these two—created an 'open-ended comment' would have returned the result `nil`, while a properly closed comment would give the result `\unique`.

Here is the file in its entirety:

```
0 entries were missing from file 1
----------------------------------

0 entries were missing from file 2
----------------------------------

37 entries were different between files
---------------------------------------
/**/: nil:Nil vs unique:Symbol
/*/ : unique:Symbol vs nil:Nil
/*/!: unique:Symbol vs nil:Nil
/*/": unique:Symbol vs nil:Nil
/*/#: unique:Symbol vs nil:Nil
/*/$: unique:Symbol vs nil:Nil
/*/%: unique:Symbol vs nil:Nil
/*/&: unique:Symbol vs nil:Nil
/*/': unique:Symbol vs nil:Nil
/*/(: unique:Symbol vs nil:Nil
/*/): unique:Symbol vs nil:Nil
/*/+: unique:Symbol vs nil:Nil
/*/,: unique:Symbol vs nil:Nil
/*/-: unique:Symbol vs nil:Nil
/*/.: unique:Symbol vs nil:Nil
/*/0: unique:Symbol vs nil:Nil
/*/9: unique:Symbol vs nil:Nil
/*/:: unique:Symbol vs nil:Nil
/*/;: unique:Symbol vs nil:Nil
/*/<: unique:Symbol vs nil:Nil
/*/=: unique:Symbol vs nil:Nil
/*/>: unique:Symbol vs nil:Nil
/*/?: unique:Symbol vs nil:Nil
/*/@: unique:Symbol vs nil:Nil
/*/A: unique:Symbol vs nil:Nil
/*/Z: unique:Symbol vs nil:Nil
/*/[: unique:Symbol vs nil:Nil
/*/\: unique:Symbol vs nil:Nil
/*/]: unique:Symbol vs nil:Nil
/*/_: unique:Symbol vs nil:Nil
/*/`: unique:Symbol vs nil:Nil
/*/a: unique:Symbol vs nil:Nil
/*/z: unique:Symbol vs nil:Nil
/*/{: unique:Symbol vs nil:Nil
/*/|: unique:Symbol vs nil:Nil
/*/}: unique:Symbol vs nil:Nil
/*/~: unique:Symbol vs nil:Nil
```

This is exactly what we'd expect—the string `/*/**/*/` makes a properly closed nested block comment, and strings like `/*/*/a*/` make an open-ended nested block comment, whereas before the last `/` of the initial `/*/*/` would have closed the inner block comment. And, we've verified in the process that absolutely none of the other behavior tested by this suite (covering over 1 billion unique, mostly ill-formed strings) has changed.

**Comment by brianlheim** [comment]
To clarify—when I say "merged in what I could" above, I just mean that I only cherry-picked the head commit; the penultimate commit created a merge conflict that I didn't really want to deal with, plus it still significantly reduced the volume of parser error and info messages.

File: testsuite/sclang/LexerParserCompilerTestUtils.sc (hunk 2 above)
Comment (brianlheim) [inline testsuite/sclang/LexerParserCompilerTestUtils.sc]: Just wanted to leave a note that I've changed this :)

File: testsuite/sclang/TestLexerTargetedBrutal.sc (hunk 3 above)
Comment (brianlheim) [inline testsuite/sclang/TestLexerTargetedBrutal.sc]: Fair; speed isn't a concern here though, and I figured using an IdentityDict might imply something more clever is happening than is really the case. :) Do you mind if I leave it?

File: testsuite/sclang/lpc/README_LPC_TESTS.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# How to run these tests
+
+1. Unzip test_output.zip, and place its four folders in this directory.
```
Comment (snappizz) [inline testsuite/sclang/lpc/README_LPC_TESTS.md]: Please add a .gitignore for the unzipped files.

**Comment by snappizz** [comment]
I am working on a thorough review of this PR. Please do not merge until I'm done, thanks!

**Comment by snappizz** [comment]
So far I've only run `TestLexerBrutal`, and some of the tests are failing. (I'm working on a branch that merges this PR + special-silent-sc + master)

File: testsuite/sclang/lpc/README_LPC_TESTS.md (hunk 4 above)
Comment (brianlheim) [inline testsuite/sclang/lpc/README_LPC_TESTS.md]: Done, and also updated the readme about that issue with the class lib/MathLib.

**Comment by snappizz** [comment]
Diff in `mini_4_basic` :sob: 

```
0 entries were missing from file 1
----------------------------------

0 entries were missing from file 2
----------------------------------

10 entries were different between files
---------------------------------------
 0/0: -nan:Float vs nan:Float
-0/0: -nan:Float vs nan:Float
0 /0: -nan:Float vs nan:Float
0/ 0: -nan:Float vs nan:Float
0/0 : -nan:Float vs nan:Float
0/0": -nan:Float vs nan:Float
0/0': -nan:Float vs nan:Float
0/00: -nan:Float vs nan:Float
0/0;: -nan:Float vs nan:Float
00/0: -nan:Float vs nan:Float
```

**Comment by snappizz** [comment]
Also, `LID` is causing more parser differences.

**Comment by brianlheim** [comment]
`-nan` is a bug value and we should open an issue for it. nan is _not a number_, you cannot have _negative not a number_.

I didn't think of `LID`... What do you think the easiest workaround is there? I can think of a few solutions:

- Force manual disable of LID: add a check in `AbstractLPCBrutalTest` that checks if a class `LID` is loaded and if it is, throw an error with a message. Also document elsewhere that on Linux this class has to be removed. Bad because it requires manual workaround, but at least conforms in theory.
- Give the interpreter a preProcessor function that replaces any string `LID` with `LHD` or something like that. Bad because it interferes directly with testing.
- Do nothing about failures and leave a note that this is OK. Bad because it requires manual verification/workaround.
- Don't have the lexer test with a configuration that could produce the string `LID`. Either exclude most capital letters form the `half` alphabet or use the `mini` alphabet for that range. Bad because it hides the problem and reduces test coverage; on the other hand, I'm not totally convinced that the `half` alphabet needs to include all capital letters. Most of the interesting strings for the lexer use lowercase letters and punctuation, and lexing/compiling of class-name strings is covered well by the parser and compiler tests.

Since the purpose of this test is just to confirm that the same strings are lexed the same way, I'd consider this a pass. The details of execution and numerical representation are not part of the lexer's domain.

**Comment by snappizz** [comment]
I'd vouch for forbidding the string `LID`. It (very marginally) reduces the number of test cases.

**Comment by jamshark70** [comment]
> -nan is a bug value and we should open an issue for it. nan is not a number, you cannot have negative not a number.

`-nan` exists in IEEE floating point land. The sign bit is independent of the value: mathematically, it is impossible for 0 to have a sign, but IEEE floats allow -0... hence, `-0/0` is a thing and, if I'm not mistaken, does evaluate to `-nan`.

If it's a bug value, it isn't our bug. We could legitimately produce a parser error for the token `-nan` but still, the statement "you cannot have negative not a number" is not exactly right.

**Comment by brianlheim** [comment]
> but still, the statement "you cannot have negative not a number" is not exactly right.

Yeah, I was thinking purely of the mathematical definition. I guess it so happens that my particular hardware+toolchain always results in a `nan` IEEE floating point representation with a 0 sign bit. I compiled a test program using GCC instead of AppleClang out of curiosity, and suddenly the sign bits are set with `0.0f/0.0f` and `-0.0f/0.0f`. However, `printf` still prints `nan` and not `-nan` for those values, so it looks like my behavior is a combination of compiler behavior and vendor headers.

> If it's a bug value, it isn't our bug.

To be fair, this could be defined in SC; we're not completely bound to the behavior of C or any other language, and we are given the `isnan` macro.

> We could legitimately produce a parser error for the token -nan

Actually, right now `nan` isn't even recognized as a legal string in SC, although `inf` is. I've wondered about changing this.

**Comment by brianlheim** [comment]
So, IMO the best place to add these workarounds is in [`doOutputsMatch`](https://github.com/supercollider/supercollider/pull/2751/files#diff-c4d3f6ae9d958d62b8d60145f0bafe3aR559)—the outputs would be considered matching if the class of either was `LID` or `Meta_LID`, and the outputs would also be considered matching if the class of both was `Float` and the string representation of both ended with `"nan"`. This makes more sense to me than hacking the evaluation and writing mechanism; it's sort of the same as asking "are the results equal within tolerance"?

**Comment by snappizz** [comment]
Up to you. I'm not going to be perfectionistic here. As long as the tests pass, it's fine by me :smile:

**Comment by telephon** [comment]
I'm also fine any way.

>Yeah, I was thinking purely of the mathematical definition

It is a disputed question whether a formalism should completely fail when you have a value in it that is a signal for not valid in the formalism. A sign may carry some information that could indicate how to continue. 

But that aside, it seems unusual to have more than one bottom value in a language semantics (https://en.wikipedia.org/wiki/Bottom_type).

**Comment by brianlheim** [comment]
> it seems unusual to have more than one bottom value

I think there's a misunderstanding here, and I have to credit James for pointing it out. The actual _value_ of [-]NaN is something that behaves the same everywhere. In other words, a value that prints as `nan` and a value that prints as `-nan` are interchangeable and equivalent.

> A sign (e.g.) may carry some information that could indicate how to continue despite the failure.

While I'd like this to be true, I've already demonstrated (and James has already pointed out) that the status of the sign bit in a NaN representation is undefined. That simply isn't information we can rely on.

**Comment by telephon** [comment]
Ah ok thank you - sorry for the noise.

**Comment by brianlheim** [comment]
@telephon No worries! Thank you again for all your input on this.

@snappizz Ready for testing. I still need to spend some time on the documentation but the tests are in good condition now. See `doOutputsMatch` for how I fixed the LID/NaN issues. I also added coverage tests to `TestLPCTestUtils` for those cases.

**Comment by snappizz** [comment]
One failure left in `TestLexerBrutal`:

    LID: "LID:Meta_LID" vs "compile-error"

**Comment by brianlheim** [comment]
@snappizz I've patched it and added a test, thanks for catching. I should have had a test for that in the first place.

**Comment by snappizz** [comment]
Failures in `TestLexerTargetedBrutal`, `alnum_3_radix36_diff`:

```

0 entries were missing from the test file
-----------------------------------------

0 entries were missing from the validation file
-----------------------------------------------

5 entries were different (test vs validation)
---------------------------------------------
.Q8: "0.72839506172839:Float" vs "0.7283950617284:Float"
.SH: "0.79089506172839:Float" vs "0.7908950617284:Float"
.UQ: "0.85339506172839:Float" vs "0.8533950617284:Float"
.WZ: "0.91589506172839:Float" vs "0.9158950617284:Float"
.Z8: "0.97839506172839:Float" vs "0.9783950617284:Float"
```

**Comment by snappizz** [comment]
I already mentioned this, but `TestParser` succeeds, and `TestCompilerBrutal` has two failures related to the presence of LID. So, all that remains to get the tests to pass are LID and the radix issue.

**Comment by snappizz** [comment]
I think it would be a good idea to add some comments elsewhere in the sclang source reminding future developers that these tests exist. I suggest advertising it to `yylex()` and the Bison source file at least.

**Comment by brianlheim** [comment]
@snappizz I've added your suggestion to the main checklist at the top of this PR. I will get around to the documentation after tests pass & your code review if that's OK.

Ready to test again. I added more regression tests to `TestLPCTestUtils` as I solved each issue. It's still possible I missed something simple. In any case, you'll now get debugging messages when certain results are skipped for `nan`, `LID`, or floating-point precision reasons, which I think is a good thing to have. As the responsibilities of `doOutputsMatch` increase, I worry about it over-accepting results.

Also, per discussion on Gitter, this PR should be squash-merged. I'm a little sad that all my nice commit messages will disappear, but I think the code is well-commented where important or strange decisions had to be made.

**Comment by snappizz** [comment]
@brianlheim

> I'm a little sad that all my nice commit messages will disappear, but I think the code is well-commented where important or strange decisions had to be made.

You can rebase and still preserve some commit messages as desired.

```
~/git/supercollider/testsuite/sclang/lpc [sclang-tests L|✚ 1⚑ 2] 
17:32 $ tail -n +1 */*_diff
==> compiler/allChars_3_basicNoTCO_diff <==

0 entries were missing from the test file
-----------------------------------------

0 entries were missing from the validation file
-----------------------------------------------

1 entries were different (test vs validation)
---------------------------------------------
LID: "0000F2" vs "compile-error"

==> compiler/allChars_3_basicTCO_diff <==

0 entries were missing from the test file
-----------------------------------------

0 entries were missing from the validation file
-----------------------------------------------

1 entries were different (test vs validation)
---------------------------------------------
LID: "0000F2" vs "compile-error"

==> lexer/half_3_basic_diff <==

0 entries were missing from the test file
-----------------------------------------

0 entries were missing from the validation file
-----------------------------------------------

1 entries were different (test vs validation)
---------------------------------------------
LID: "LID:Meta_LID" vs "compile-error"

==> lexer/half_3_semanticPrefix_diff <==

0 entries were missing from the test file
-----------------------------------------

0 entries were missing from the validation file
-----------------------------------------------

1 entries were different (test vs validation)
---------------------------------------------
LID: "LID:Meta_LID" vs "compile-error"

==> lexer/half_3_semanticSuffix_diff <==

0 entries were missing from the test file
-----------------------------------------

0 entries were missing from the validation file
-----------------------------------------------

1 entries were different (test vs validation)
---------------------------------------------
LID: "unique:Symbol" vs "compile-error"
```

**Comment by brianlheim** [comment]
OK, I defined a dummy LID class on my own machine and made sure it works now. You should see a couple lines like this while running:

```
[debug] LPCTestUtils: Ignoring a result because of LID class.	Input: [ L, I, D ]	Output 1: [ 4C4944, Meta_LID ]	Output 2: [ !cErr ]
```

I'll try to run these tests on Windows later today.

**Comment by snappizz** [comment]
All the tests pass now! I'll do code review over the next few days, and also see what happens if I mess with `yylex()`.

**Comment by brianlheim** [comment]
These tests now also pass on Windows. (Win10+VS 2013 build)

**Comment by brianlheim** [comment]
LPCTestUtils has gotten pretty large (~900 loc). I could break out some of the file utilities (and corresponding tests) into a separate class...

**Comment by brianlheim** [comment]
Wooooo. Yah I see one that I didn't intend. The rest are for alignment unless I'm missing something obvious. OK, will document and have ready for (**squash**) merge soon.

**Comment by LFSaw** [comment]
((( just want to say thanks for the massive work you're investing into this... sorry for the noise amd keep it up, massively appreciated! :) )))

**Comment by brianlheim** [comment]
OK, done with the readme. Since this has two approved reviews and doesn't do anything with the main lib, I'm assuming this is good to merge once the CI tests pass.

To reduce noise in the commit log, eliminate the snapshots that had all those text files I committed in the first stages of this project, and avoid a rebase, I'm going to merge this as a squashed commit. Thanks again everyone for your time and reviews!
